### PR TITLE
Redesign handshake hashs

### PR DESCRIPTION
**Pull request by boaks**
Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
May be sometime in the future this also simplifies [RFC7627](https://tools.ietf.org/html/rfc7627)

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHandshaker.java [hunk 1]
```diff
@@ -223,11 +208,13 @@ protected void doProcessMessage(HandshakeMessage message) throws HandshakeExcept
 						String.format("Unsupported key exchange algorithm %s", getKeyExchangeAlgorithm().name()),
 						new AlertMessage(AlertLevel.FATAL, AlertDescription.HANDSHAKE_FAILURE, message.getPeer()));
 			}
+			handshakeMessages.add(message);
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHandshaker.java]: ~(clearly a details)
I feel this is a bit strange that sometime we call `handshakeMessages.add(message)` in the generic `doProcessMessage` and sometime in the specific processing method (like `receivedServerHello`).~

Fixed in next commit.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/CipherSuite.java [hunk 2]
```diff
@@ -743,12 +745,38 @@ public boolean isSupported() {
...
+		 * 
+		 * Calls {@link Mac#reset()} on access.
+		 * 
+		 * @return mac, or {@code null}, if not supported by vm.
+		 */
 		public Mac getMac() {
-			return mac == null ? null : mac.current();
+			if (mac != null) {
+				Mac current = mac.current();
+				current.reset();
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/CipherSuite.java]: was it a bug ? (the missing reset ?)

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHandshaker.java [hunk 3]
```diff
@@ -207,7 +192,7 @@ protected void doProcessMessage(HandshakeMessage message) throws HandshakeExcept
 				break;
 
 			case PSK:
-				serverKeyExchange = (PSKServerKeyExchange) message;
+//				serverKeyExchange = (PSKServerKeyExchange) message;
```
[the remark this answers is not in the document]
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHandshaker.java]: Uups, no.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/CipherSuite.java (hunk 2 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/cipher/CipherSuite.java]: I added it for cases, where something fails before and so the `doFinal` or `digest` is not called. But I didn't recognized, that actually this happens.

**Comment by boaks** [comment]
I squashed the commits not to change the same files over and over for the same.

**Comment by sbernard31** [comment]
Ok

**Pull request merged**